crates/types/src/predicate/header.rs

```diff
@@ -0,0 +1,598 @@
+    if bounds.num_state_reads > Predicate::MAX_STATE_READS {
+        return Err(PredicateError::TooManyStateReads(bounds.num_state_reads));
+    }
+    if bounds.num_constraints > Predicate::MAX_CONSTRAINTS {
+        return Err(PredicateError::TooManyConstraints(bounds.num_constraints));
+    }
+
+    // Count the total size of the state read programs.
+    let mut state_read_lens_sum: usize = 0;
+
+    // Check the size of each program.
+    if let Some(err) = bounds.state_read_lens.find_map(|len| {
+        state_read_lens_sum = state_read_lens_sum.saturating_add(len);
+        (len > Predicate::MAX_STATE_READ_SIZE_BYTES)
+            .then_some(PredicateError::StateReadTooLarge(len))
+    }) {
+        return Err(err);
+    }
+
+    // Count the total size of the constraint check programs.
+    let mut constraint_lens_sum: usize = 0;
+
+    // Check the size of each program.
+    if let Some(err) = bounds.constraint_lens.find_map(|len| {
+        constraint_lens_sum = constraint_lens_sum.saturating_add(len);
+        (len > Predicate::MAX_CONSTRAINT_SIZE_BYTES)
+            .then_some(PredicateError::ConstraintTooLarge(len))
+    }) {
+        return Err(err);
+    }
+
+    // Calculate the total encoded size of the predicate.
+    let encoded_size = encoded_size(&EncodedSize {
+        num_state_reads: bounds.num_state_reads,
+        num_constraints: bounds.num_constraints,
+        state_read_lens_sum,
+        constraint_lens_sum,
+    });
+
+    // Check the total size of the encoded predicate.
+    if encoded_size > Predicate::MAX_BYTES {
+        return Err(PredicateError::PredicateTooLarge(encoded_size));
+    }
+
+    Ok(())
+}
+
+impl TryFrom<&Predicate> for EncodedHeader {
+    type Error = PredicateError;
+
+    /// Creates the encoded header from a [`Predicate`].
+    fn try_from(predicate: &Predicate) -> Result<Self, Self::Error> {
+        let static_header = EncodedFixedSizeHeader::try_from(predicate)?;
+        let lens = encode_program_lengths(predicate);
+        Ok(Self {
+            fixed_size_header: static_header,
+            lens,
+        })
+    }
+}
+
+impl TryFrom<&Predicate> for EncodedFixedSizeHeader {
+    type Error = PredicateError;
+
+    fn try_from(predicate: &Predicate) -> Result<Self, Self::Error> {
+        Ok(EncodedFixedSizeHeader::from(FixedSizeHeader::try_from(
+            predicate,
+        )?))
+    }
+}
+
+impl TryFrom<&Predicate> for FixedSizeHeader {
+    type Error = PredicateError;
+
+    fn try_from(predicate: &Predicate) -> Result<Self, Self::Error> {
+        predicate.check_predicate_bounds()?;
+
+        Ok(Self {
+            num_state_reads: predicate.state_read.len() as u8,
+            num_constraints: predicate.constraints.len() as u8,
+        })
+    }
+}
+
+/// Encode the lengths of the [`Predicate::state_read`]
+/// and [`Predicate::constraints`] programs into bytes.
+///
+/// Lengths are encoded as big-endian [`u16`] bytes.
```

Comment:
Just crossed my mind - maybe we should consider making these `u32` for future-proofness?

It doesn't seem infeasible to me that someone one day we might wanna support a single program with more than 65kb :thinking: I know that we have a fixed limit on bytecode length at the moment anyway, but that's just a constant that can be changed in the future if we decide to increase it. This on the other hand would be much harder to change as its baked into the way we calculate predicate addresses :thinking:

Reply:
Yeh it's a hard one because it would double the bytes used for every length which would have a real impact on network bandwidth. It's hard to know how big pint programs will be but 65k ops seems like a lot and you can always split a constraint into 2 if it doesn't fit